integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py

```diff
@@ -363,50 +350,47 @@ def _normalize_filters(filters: Dict[str, Any]) -> Tuple[List[str], Dict[str, An
             raise ChromaDocumentStoreFilterError(msg)
 
         ids = []
         where = defaultdict(list)
-        where_document = defaultdict(list)
-        keys_to_remove = []
+        where_document = {}
+        document_flag = False
 
-        for field, value in filters.items():
+        # Process filters
+        for field, value in list(filters.items()):
             if field == "content":
-                # Schedule for removal the original key, we're going to change it
-                keys_to_remove.append(field)
                 where_document["$contains"] = value
+                document_flag = True
+                del filters[field]
             elif field == "id":
-                # Schedule for removal the original key, we're going to change it
-                keys_to_remove.append(field)
                 ids.append(value)
-            elif isinstance(value, (list, tuple)):
-                # Schedule for removal the original key, we're going to change it
-                keys_to_remove.append(field)
-
-                # if the list is empty the filter is invalid, let's just remove it
-                if len(value) == 0:
+                del filters[field]
+            elif isinstance(value, (list, tuple)) and field not in ["$and", "$or"]:
+                if not value:  # Skip empty lists
+                    del filters[field]
                     continue
 
-                # if the list has a single item, just make it a regular key:value filter pair
                 if len(value) == 1:
-                    where[field] = value[0]
-                    continue
-
-                # if the list contains multiple items, we need an $or chain
-                for v in value:
-                    where["$or"].append({field: v})
-
-        for k in keys_to_remove:
-            del filters[k]
-
-        final_where = dict(filters)
-        final_where.update(dict(where))
+                    where[field] = value[0]  # Simplify single-item lists
+                # Check if list contains special operators that apply to documents
+                elif any(isinstance(v, dict) and any(k in v for k in ["$contains", "$not_contains"]) for v in value):
+                    where_document = filters.copy()
+                    document_flag = True
+                    break
+                else:
+                    where[field] = {"$in": value}  # Create $in chain for multiple items
+                del filters[field]
+
+        final_where = {}
         try:
-            if final_where:
-                validate_where(final_where)
-            if where_document:
+            if document_flag:
                 validate_where_document(where_document)
+            elif filters or where:
+                final_where.update(filters)
+                final_where.update(where)
+                validate_where(final_where)
         except ValueError as e:
             raise ChromaDocumentStoreFilterError(e) from e
```

Comment:
I think we can provide some extra context here, i.e., mention what operation raise the error.

Reply:
The error message generated by the internal chroma function usually highlights which part of the filter is problematic. So I think this should be sufficient.

Reply:
Sure, but we should prefix that with something along the lines of `Failed to validate WhereDocument` (unless the library mentions that too).